contracts/handlers/ERC20Handler.sol

```diff
@@ -92,24 +92,113 @@ contract ERC20Handler is IDepositHandler, ERC20Safe {
+    function parseInt(string memory _a, uint _b) internal pure returns (uint _parsedInt) {
+        bytes memory bresult = bytes(_a);
+        uint mint = 0;
+        bool decimals = false;
+        for (uint i = 0; i < bresult.length; i++) {
+            if ((uint(uint8(bresult[i])) >= 48) && (uint(uint8(bresult[i])) <= 57)) {
+                if (decimals) {
+                   if (_b == 0) {
+                       break;
+                   } else {
+                       _b--;
+                   }
+                }
+                mint *= 10;
+                mint += uint(uint8(bresult[i])) - 48;
+            } else if (uint(uint8(bresult[i])) == 46) {
+                decimals = true;
+            }
+        }
+        if (_b > 0) {
+            mint *= 10 ** _b;
+        }
+        return mint;
+    }
+
+    // copied from https://github.com/provable-things/ethereum-api/blob/master/oraclizeAPI_0.5.sol#L872
+    function parseAddr(string memory _a) internal pure returns (address _parsedAddress) {
+        bytes memory tmp = bytes(_a);
+        uint160 iaddr = 0;
+        uint160 b1;
+        uint160 b2;
+        for (uint i = 2; i < 2 + 2 * 20; i += 2) {
+            iaddr *= 256;
+            b1 = uint160(uint8(tmp[i]));
+            b2 = uint160(uint8(tmp[i + 1]));
+            if ((b1 >= 97) && (b1 <= 102)) {
+                b1 -= 87;
+            } else if ((b1 >= 65) && (b1 <= 70)) {
+                b1 -= 55;
+            } else if ((b1 >= 48) && (b1 <= 57)) {
+                b1 -= 48;
+            }
+            if ((b2 >= 97) && (b2 <= 102)) {
+                b2 -= 87;
+            } else if ((b2 >= 65) && (b2 <= 70)) {
+                b2 -= 55;
+            } else if ((b2 >= 48) && (b2 <= 57)) {
+                b2 -= 48;
+            }
+            iaddr += (b1 * 16 + b2);
+        }
+        return address(iaddr);
+    }
+
+    // copied from https://ethereum.stackexchange.com/questions/31457/substring-in-solidity
+    function subString(string memory str, uint startIndex, uint endIndex) internal pure returns (string memory) {
+        bytes memory strBytes = bytes(str);
+        bytes memory result = new bytes(endIndex-startIndex);
+        for(uint i = startIndex; i < endIndex; i++) {
+            result[i-startIndex] = strBytes[i];
+        }
+        return string(result);
+    }
+
     function executeDeposit(bytes memory data) public override _onlyBridge {
-        address destinationChainTokenAddress;
-        address destinationRecipientAddress;
-        uint256 amount;
+        string memory tokenID;
+        address       destinationRecipientAddress;
+        uint256       amount;
 
         assembly {
-            destinationChainTokenAddress := mload(add(data, 0x20))
-            destinationRecipientAddress  := mload(add(data, 0x40))
-            amount                       := mload(add(data, 0x60))
+            tokenID                     := mload(add(data, 0x20))
```

Comment:
so since a string is variable length we can't load it as the 32 bytes after address 0x20 in the data, as we don't know if it's exactly 32 bytes long. Best to have the `tokenID` at the end of `data` so we can use a similar approach to calldatacopy it out of the calldata.

Reply:
whoops didn't realise this was a draft, should have just left a comment